src/softmax.jl

```diff
@@ -52,4 +17,69 @@ independent.
       0.244728
       0.665241
 """
-softmax
+softmax(xs) = softmax!(similar(xs), xs)
+
+function softmax!(out::AbstractVecOrMat{T}, xs::AbstractVecOrMat{T}) where {T}
```

Comment:
Is it possible (without hurting performance) to make this generic over the dimension index? (related: it would be pretty nice if using the clean broadcast versions in the comments didn't hurt performance either—is that because the broadcast infrastructure doesn't insert the equivalent of `@inbounds`?)

Reply:
In my experience, broadcasting creates small objects, and so if you're doing lots of small broadcasts, you can easily get drowned in small allocations.  It's almost always faster to just write the loop yourself, unfortunately.

Reply:
I'm going to punt any improvements here off onto https://github.com/FluxML/NNlib.jl/issues/77; this PR is big enough as it is.  ;)